snapshot/snapshotfs/snapshot_tree_walker.go

```diff
@@ -106,21 +108,9 @@ func (w *TreeWalker) processDirEntry(ctx context.Context, dir fs.Directory, entr
 	var ag workshare.AsyncGroup
 	defer ag.Wait()
 
-	entries, err := dir.Readdir(ctx)
-	if err != nil {
-		w.ReportError(ctx, entryPath, errors.Wrap(err, "error reading directory"))
-		return
-	}
-
-	for _, ent := range entries {
-		ent := ent
-
-		if w.TooManyErrors() {
-			break
-		}
-
-		if w.alreadyProcessed(ent) {
-			continue
+	err := dir.IterateEntries(ctx, func(c context.Context, ent fs.Entry) error {
+		if w.TooManyErrors() || w.alreadyProcessed(ent) {
+			return nil
```

Comment:
nit: may be worth doing the same trick with returning dummy error and treating it as success to stop the iteration early in case of TooManyErrors(). Probably does not matter much.